File: app/src/main/java/com/uccd3223/p1_chai_boon_hong_2206806/SequenceActivity.java
package com.uccd3223.p1_chai_boon_hong_2206806;

import android.content.ClipData;
import android.content.ClipDescription;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.view.DragEvent;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.core.content.ContextCompat;

import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil;
import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil.SortDirection;
import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil.SortExercise;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

public class SequenceActivity extends BaseGameActivity {
    private interface BalloonViewFactory {
        FrameLayout create(int index, LinearLayout row);
    }

    private static final int EMPTY_POSITION = -1;
    private static final String STATE_OPTIONS = "sequence_options";
    private static final String STATE_DIRECTION = "sequence_direction";
    private static final String STATE_PLACED = "sequence_placed";
    private static final String STATE_LAST_SIGNATURE = "sequence_last_signature";
    private static final String STATE_QUESTION_LOCKED = "sequence_question_locked";

    private LinearLayout sequenceContainerLayout;
    private LinearLayout optionsContainerLayout;
    private TextView tvStartDirection;
    private TextView tvFeedback;
    private SortExercise currentExercise;
    private List<Integer> placedNumbers = new ArrayList<>();
    private String lastSignature = "";
    private boolean questionLocked;
    private final List<FrameLayout> targetViews = new ArrayList<>();
    private final Map<Integer, FrameLayout> sourceViews = new HashMap<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sequence);
        setupGameModeUI();
        initializeUI();
        if (savedInstanceState != null && savedInstanceState.containsKey(STATE_OPTIONS)) {
            restoreQuestion(savedInstanceState);
        } else {
            loadNextQuestion();
        }
    }

    private void initializeUI() {
        sequenceContainerLayout = findViewById(R.id.sequenceContainerLayout);
        optionsContainerLayout = findViewById(R.id.optionsContainerLayout);
        tvStartDirection = findViewById(R.id.tvStartDirection);
        tvFeedback = findViewById(R.id.tvFeedback);
    }

    private void loadNextQuestion() {
        questionLocked = false;
        resetFeedback(tvFeedback);
        do {
            int length = ThreadLocalRandom.current().nextInt(3, 7);
            currentExercise = ExerciseGeneratorUtil.generateSortExercise(length);
        } while (questionSignature(currentExercise).equals(lastSignature));
        lastSignature = questionSignature(currentExercise);
        placedNumbers = new ArrayList<>(Collections.nCopies(currentExercise.getOptions().size(), null));
        renderQuestion();
    }

    private void restoreQuestion(Bundle state) {
        int[] optionArray = state.getIntArray(STATE_OPTIONS);
        List<Integer> options = new ArrayList<>();
        if (optionArray != null) {
            for (int option : optionArray) options.add(option);
        }
        if (options.size() < 3 || options.size() > 6
                || new HashSet<>(options).size() != options.size()) {
            loadNextQuestion();
            return;
        }

        int directionIndex = Math.max(
                0,
                Math.min(SortDirection.values().length - 1, state.getInt(STATE_DIRECTION))
        );
        currentExercise = new SortExercise(options, SortDirection.values()[directionIndex]);
        int[] restoredArray = state.getIntArray(STATE_PLACED);
        List<Integer> restored = new ArrayList<>();
        if (restoredArray != null) {
            for (int value : restoredArray) restored.add(value == EMPTY_POSITION ? null : value);
        }
        placedNumbers = restored.size() == options.size()
                ? restored
                : new ArrayList<>(Collections.nCopies(options.size(), null));
        String restoredSignature = state.getString(STATE_LAST_SIGNATURE);
        lastSignature = restoredSignature == null ? "" : restoredSignature;
        questionLocked = state.getBoolean(STATE_QUESTION_LOCKED);
        renderQuestion();
        if (questionLocked) {
            restoreCompletedQuestion(tvFeedback, successMessage(), this::loadNextQuestion);
        }
    }

    private String questionSignature(SortExercise exercise) {
        List<Integer> sorted = exercise.getOptions();
        Collections.sort(sorted);
        StringBuilder signature = new StringBuilder(exercise.getDirection().name()).append(':');
        for (int index = 0; index < sorted.size(); index++) {
            if (index > 0) signature.append(',');
            signature.append(sorted.get(index));
        }
        return signature.toString();
    }

    private void renderQuestion() {
        tvStartDirection.setText(currentExercise.getDirection() == SortDirection.ASCENDING
                ? R.string.start_with_smallest
                : R.string.start_with_biggest);
        renderSequenceTargets();
        renderAnswerBalloons();
    }

    private void renderSequenceTargets() {
        targetViews.clear();
        renderBalloonRows(
                sequenceContainerLayout,
                currentExercise.getOptions().size(),
                (index, row) -> {
                    BalloonFrameLayout targetView = inflateBalloon(row);
                    targetViews.add(targetView);
                    Integer placed = placedNumbers.get(index);
                    if (placed == null) {
                        targetView.<TextView>findViewById(R.id.balloonText).setText("");
                        targetView.setContentDescription(getString(
                                R.string.position_empty,
                                index + 1,
                                currentExercise.getOptions().size()
                        ));
                        attachDropTarget(targetView, index);
                    } else {
                        fillTargetView(targetView, index, placed);
                    }
                    return targetView;
                }
        );
    }

    private void renderAnswerBalloons() {
        sourceViews.clear();
        List<Integer> options = currentExercise.getOptions();
        renderBalloonRows(
                optionsContainerLayout,
                options.size(),
                (index, row) -> {
                    int number = options.get(index);
                    BalloonFrameLayout optionView = inflateBalloon(row);
                    sourceViews.put(number, optionView);
                    optionView.<TextView>findViewById(R.id.balloonText).setText(
                            getString(R.string.number_value, number)
                    );
                    optionView.<ImageView>findViewById(R.id.balloonBg).setImageTintList(
                            ColorStateList.valueOf(colorForNumber(number))
                    );
                    optionView.setContentDescription(getString(R.string.drag_number, number));
                    optionView.setVisibility(placedNumbers.contains(number) ? View.INVISIBLE : View.VISIBLE);
                    optionView.setOnClickListener(view -> placeByTap(number, optionView));
                    attachImmediateDrag(optionView, number);
                    return optionView;
                }
        );
    }

    private void renderBalloonRows(
            LinearLayout parent,
            int count,
            BalloonViewFactory factory
    ) {
        parent.removeAllViews();
        List<Integer> pattern = ExerciseGeneratorUtil.balloonRowPattern(count);
        int itemIndex = 0;
        for (int rowSize : pattern) {
            LinearLayout row = new LinearLayout(this);
            row.setOrientation(LinearLayout.HORIZONTAL);
            row.setGravity(Gravity.CENTER);
            row.setLayoutParams(new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
            ));
            for (int item = 0; item < rowSize; item++) {
                row.addView(factory.create(itemIndex++, row));
            }
            parent.addView(row);
        }
    }

    private BalloonFrameLayout inflateBalloon(LinearLayout parent) {
        int size = getResources().getDimensionPixelSize(R.dimen.balloon_size);
        int gap = getResources().getDimensionPixelSize(R.dimen.balloon_gap);
        BalloonFrameLayout balloon = (BalloonFrameLayout) getLayoutInflater().inflate(
                R.layout.item_balloon,
                parent,
                false
        );
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(size, size);
        params.setMargins(gap, gap, gap, gap);
        balloon.setLayoutParams(params);
        return balloon;
    }

    private void attachDropTarget(FrameLayout targetView, int targetIndex) {
        targetView.setOnDragListener((view, event) -> {
            View source = event.getLocalState() instanceof View
                    ? (View) event.getLocalState()
                    : null;
            switch (event.getAction()) {
                case DragEvent.ACTION_DRAG_STARTED:
                    return event.getClipDescription() != null
                            && event.getClipDescription().hasMimeType(ClipDescription.MIMETYPE_TEXT_PLAIN)
                            && source != null;
                case DragEvent.ACTION_DRAG_ENTERED:
                    view.setAlpha(0.7f);
                    return true;
                case DragEvent.ACTION_DRAG_EXITED:
                    view.setAlpha(1f);
                    return true;
                case DragEvent.ACTION_DROP:
                    view.setAlpha(1f);
                    Integer number = parseDraggedNumber(event);
                    if (number != null
                            && source != null
                            && placedNumbers.get(targetIndex) == null
                            && number.equals(currentExercise.getOrderedNumbers().get(targetIndex))) {
                        placeNumber(targetIndex, number, source);
                    } else {
                        showSequenceRetry(view);
                    }
                    return true;
                case DragEvent.ACTION_DRAG_ENDED:
                    view.setAlpha(1f);
                    if (source != null) source.setAlpha(1f);
                    return true;
                default:
                    return false;
            }
        });
    }

    private Integer parseDraggedNumber(DragEvent event) {
        if (event.getClipData() == null || event.getClipData().getItemCount() == 0) return null;
        try {
            return Integer.valueOf(event.getClipData().getItemAt(0).getText().toString());
        } catch (RuntimeException exception) {
            return null;
        }
    }

    private void attachImmediateDrag(BalloonFrameLayout optionView, int number) {
        int touchSlop = ViewConfiguration.get(this).getScaledTouchSlop();
        float[] down = new float[2];
        boolean[] dragStarted = new boolean[1];

        optionView.setOnTouchListener((view, event) -> {
            switch (event.getActionMasked()) {
                case MotionEvent.ACTION_DOWN:
                    down[0] = event.getX();
                    down[1] = event.getY();
                    dragStarted[0] = false;
                    return true;
                case MotionEvent.ACTION_MOVE:
                    if (!dragStarted[0]
                            && (Math.abs(event.getX() - down[0]) > touchSlop
                            || Math.abs(event.getY() - down[1]) > touchSlop)) {
                        ClipData dragData = new ClipData(
                                Integer.toString(number),
                                new String[] { ClipDescription.MIMETYPE_TEXT_PLAIN },
                                new ClipData.Item(Integer.toString(number))
                        );
                        dragStarted[0] = view.startDragAndDrop(
                                dragData,
                                new View.DragShadowBuilder(view),
                                view,
                                0
                        );
                        if (dragStarted[0]) {
                            view.setAlpha(0.35f);
                            view.getParent().requestDisallowInterceptTouchEvent(true);
                        }
                    }
                    return dragStarted[0];
                case MotionEvent.ACTION_UP:
                    view.getParent().requestDisallowInterceptTouchEvent(false);
                    if (!dragStarted[0]) view.performClick();
                    return true;
                case MotionEvent.ACTION_CANCEL:
                    view.getParent().requestDisallowInterceptTouchEvent(false);
                    view.setAlpha(1f);
                    return true;
                default:
                    return dragStarted[0];
            }
        });
    }

    private void placeByTap(int number, View sourceView) {
        if (questionLocked || !canAcceptInput() || sourceView.getVisibility() != View.VISIBLE) return;
        int nextIndex = -1;
        for (int index = 0; index < placedNumbers.size(); index++) {
            if (placedNumbers.get(index) == null) {
                nextIndex = index;
                break;
            }
        }
        if (nextIndex >= 0 && number == currentExercise.getOrderedNumbers().get(nextIndex)) {
            placeNumber(nextIndex, number, sourceView);
        } else {
            showSequenceRetry(sourceView);
        }
    }

    private void placeNumber(int targetIndex, int number, View sourceView) {
        if (questionLocked || !canAcceptInput()
                || placedNumbers.get(targetIndex) != null
                || sourceView.getVisibility() != View.VISIBLE) return;
        placedNumbers.set(targetIndex, number);
        fillTargetView(targetViews.get(targetIndex), targetIndex, number);
        sourceView.setVisibility(View.INVISIBLE);

        boolean complete = true;
        for (Integer placed : placedNumbers) {
            if (placed == null) {
                complete = false;
                break;
            }
        }
        if (complete) {
            questionLocked = true;
            celebrateQuestion(
                    targetViews.get(targetIndex),
                    tvFeedback,
                    successMessage(),
                    this::loadNextQuestion
            );
        }
    }

    private String successMessage() {
        return getString(currentExercise.getDirection() == SortDirection.ASCENDING
                ? R.string.sequence_success_small
                : R.string.sequence_success_big);
    }

    private void fillTargetView(FrameLayout targetView, int index, int number) {
        targetView.setOnDragListener(null);
        targetView.<TextView>findViewById(R.id.balloonText).setText(
                getString(R.string.number_value, number)
        );
        targetView.<ImageView>findViewById(R.id.balloonBg).setImageTintList(
                ColorStateList.valueOf(colorForNumber(number))
        );
        targetView.setContentDescription(getString(
                R.string.position_filled,
                index + 1,
                currentExercise.getOptions().size(),
                number
        ));
    }

    private void showSequenceRetry(View anchor) {
        int message = currentExercise.getDirection() == SortDirection.ASCENDING
                ? R.string.sequence_retry_small
                : R.string.sequence_retry_big;
        showRetryFeedback(anchor, tvFeedback, getString(message));
    }

    private int colorForNumber(int number) {
        int[] colors = {
                R.color.game_blue,
                R.color.game_orange,
                R.color.game_purple,
                R.color.game_green,
                R.color.game_balloon_pink,
                R.color.game_star
        };
        int index = Math.max(0, currentExercise.getOptions().indexOf(number));
        return ContextCompat.getColor(this, colors[index % colors.length]);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        if (currentExercise != null) {
            List<Integer> options = currentExercise.getOptions();
            int[] optionArray = new int[options.size()];
            int[] placedArray = new int[placedNumbers.size()];
            for (int index = 0; index < options.size(); index++) {
                optionArray[index] = options.get(index);
                Integer placed = placedNumbers.get(index);
                placedArray[index] = placed == null ? EMPTY_POSITION : placed;
            }
            outState.putIntArray(STATE_OPTIONS, optionArray);
            outState.putInt(STATE_DIRECTION, currentExercise.getDirection().ordinal());
            outState.putIntArray(STATE_PLACED, placedArray);
        }
        outState.putString(STATE_LAST_SIGNATURE, lastSignature);
        outState.putBoolean(STATE_QUESTION_LOCKED, questionLocked);
        super.onSaveInstanceState(outState);
    }
}

File: app/src/main/java/com/uccd3223/p1_chai_boon_hong_2206806/SequenceActivity.kt
package com.uccd3223.p1_chai_boon_hong_2206806

import android.content.ClipData
import android.content.ClipDescription
import android.content.res.ColorStateList
import android.os.Bundle
import android.view.DragEvent
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil
import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil.SortDirection
import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil.SortExercise
import kotlin.math.abs
import kotlin.random.Random

class SequenceActivity : BaseGameActivity() {

    private lateinit var sequenceContainerLayout: LinearLayout
    private lateinit var optionsContainerLayout: LinearLayout
    private lateinit var tvStartDirection: TextView
    private lateinit var tvFeedback: TextView
    private lateinit var currentExercise: SortExercise
    private var placedNumbers = mutableListOf<Int?>()
    private var lastSignature = ""
    private var questionLocked = false
    private val targetViews = mutableListOf<FrameLayout>()
    private val sourceViews = mutableMapOf<Int, FrameLayout>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sequence)
        setupGameModeUI()
        initializeUI()

        if (savedInstanceState?.containsKey(STATE_OPTIONS) == true) {
            restoreQuestion(savedInstanceState)
        } else {
            loadNextQuestion()
        }
    }

    private fun initializeUI() {
        sequenceContainerLayout = findViewById(R.id.sequenceContainerLayout)
        optionsContainerLayout = findViewById(R.id.optionsContainerLayout)
        tvStartDirection = findViewById(R.id.tvStartDirection)
        tvFeedback = findViewById(R.id.tvFeedback)
        findViewById<Button>(R.id.btnBack).setOnClickListener { finish() }
    }

    private fun loadNextQuestion() {
        questionLocked = false
        resetFeedback(tvFeedback)
        do {
            val length = Random.nextInt(3, 7)
            currentExercise = ExerciseGeneratorUtil.generateSortExercise(length)
        } while (questionSignature(currentExercise) == lastSignature)
        lastSignature = questionSignature(currentExercise)
        placedNumbers = MutableList(currentExercise.options.size) { null }
        renderQuestion()
    }

    private fun restoreQuestion(state: Bundle) {
        val options = state.getIntArray(STATE_OPTIONS)?.toList().orEmpty()
        if (options.size !in 3..6 || options.distinct().size != options.size) {
            loadNextQuestion()
            return
        }
        val direction = SortDirection.entries[
            state.getInt(STATE_DIRECTION).coerceIn(0, SortDirection.entries.lastIndex)
        ]
        currentExercise = SortExercise(options, direction)
        val restored = state.getIntArray(STATE_PLACED)
            ?.map { value -> value.takeUnless { it == EMPTY_POSITION } }
            .orEmpty()
        placedNumbers = if (restored.size == options.size) {
            restored.toMutableList()
        } else {
            MutableList(options.size) { null }
        }
        lastSignature = state.getString(STATE_LAST_SIGNATURE).orEmpty()
        questionLocked = state.getBoolean(STATE_QUESTION_LOCKED)
        renderQuestion()
        if (questionLocked) {
            restoreCompletedQuestion(tvFeedback, successMessage(), ::loadNextQuestion)
        }
    }

    private fun questionSignature(exercise: SortExercise): String =
        "${exercise.direction}:${exercise.options.sorted().joinToString(",")}"

    private fun renderQuestion() {
        tvStartDirection.setText(
            if (currentExercise.direction == SortDirection.ASCENDING) {
                R.string.start_with_smallest
            } else {
                R.string.start_with_biggest
            }
        )
        renderSequenceTargets()
        renderAnswerBalloons()
    }

    private fun renderSequenceTargets() {
        targetViews.clear()
        renderBalloonRows(sequenceContainerLayout, currentExercise.options.size) { index, row ->
            val targetView = inflateBalloon(row)
            targetViews += targetView
            val placed = placedNumbers[index]
            if (placed == null) {
                targetView.findViewById<TextView>(R.id.balloonText).text = ""
                targetView.contentDescription = getString(
                    R.string.position_empty,
                    index + 1,
                    currentExercise.options.size
                )
                attachDropTarget(targetView, index)
            } else {
                fillTargetView(targetView, index, placed)
            }
            targetView
        }
    }

    private fun renderAnswerBalloons() {
        sourceViews.clear()
        renderBalloonRows(optionsContainerLayout, currentExercise.options.size) { index, row ->
            val number = currentExercise.options[index]
            val optionView = inflateBalloon(row)
            sourceViews[number] = optionView
            optionView.findViewById<TextView>(R.id.balloonText).text =
                getString(R.string.number_value, number)
            optionView.findViewById<ImageView>(R.id.balloonBg).imageTintList =
                ColorStateList.valueOf(colorForNumber(number))
            optionView.contentDescription = getString(R.string.drag_number, number)
            optionView.visibility = if (number in placedNumbers) View.INVISIBLE else View.VISIBLE
            optionView.setOnClickListener { placeByTap(number, optionView) }
            attachImmediateDrag(optionView, number)
            optionView
        }
    }

    private fun renderBalloonRows(
        parent: LinearLayout,
        count: Int,
        createView: (Int, LinearLayout) -> FrameLayout
    ) {
        parent.removeAllViews()
        val pattern = ExerciseGeneratorUtil.balloonRowPattern(count)
        var itemIndex = 0
        pattern.forEach { rowSize ->
            val row = LinearLayout(this).apply {
                orientation = LinearLayout.HORIZONTAL
                gravity = android.view.Gravity.CENTER
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            }
            repeat(rowSize) {
                row.addView(createView(itemIndex++, row))
            }
            parent.addView(row)
        }
    }

    private fun inflateBalloon(parent: LinearLayout): BalloonFrameLayout {
        val size = resources.getDimensionPixelSize(R.dimen.balloon_size)
        val gap = resources.getDimensionPixelSize(R.dimen.balloon_gap)
        return (layoutInflater.inflate(
            R.layout.item_balloon,
            parent,
            false
        ) as BalloonFrameLayout).apply {
            layoutParams = LinearLayout.LayoutParams(size, size).apply {
                setMargins(gap, gap, gap, gap)
            }
        }
    }

    private fun attachDropTarget(targetView: FrameLayout, targetIndex: Int) {
        targetView.setOnDragListener { view, event ->
            val source = event.localState as? View
            when (event.action) {
                DragEvent.ACTION_DRAG_STARTED ->
                    event.clipDescription.hasMimeType(ClipDescription.MIMETYPE_TEXT_PLAIN) && source != null
                DragEvent.ACTION_DRAG_ENTERED -> {
                    view.alpha = 0.7f
                    true
                }
                DragEvent.ACTION_DRAG_EXITED -> {
                    view.alpha = 1f
                    true
                }
                DragEvent.ACTION_DROP -> {
                    view.alpha = 1f
                    val number = event.clipData.getItemAt(0).text.toString().toIntOrNull()
                    if (number != null && source != null &&
                        placedNumbers[targetIndex] == null &&
                        number == currentExercise.orderedNumbers[targetIndex]
                    ) {
                        placeNumber(targetIndex, number, source)
                    } else {
                        showSequenceRetry(view)
                    }
                    true
                }
                DragEvent.ACTION_DRAG_ENDED -> {
                    view.alpha = 1f
                    source?.alpha = 1f
                    true
                }
                else -> false
            }
        }
    }

    private fun attachImmediateDrag(optionView: BalloonFrameLayout, number: Int) {
        val touchSlop = ViewConfiguration.get(this).scaledTouchSlop
        var downX = 0f
        var downY = 0f
        var dragStarted = false

        optionView.setOnTouchListener { view, event ->
            when (event.actionMasked) {
                MotionEvent.ACTION_DOWN -> {
                    downX = event.x
                    downY = event.y
                    dragStarted = false
                    true
                }
                MotionEvent.ACTION_MOVE -> {
                    if (!dragStarted &&
                        (abs(event.x - downX) > touchSlop || abs(event.y - downY) > touchSlop)
                    ) {
                        val dragData = ClipData(
                            number.toString(),
                            arrayOf(ClipDescription.MIMETYPE_TEXT_PLAIN),
                            ClipData.Item(number.toString())
                        )
                        dragStarted = view.startDragAndDrop(
                            dragData,
                            View.DragShadowBuilder(view),
                            view,
                            0
                        )
                        if (dragStarted) {
                            view.alpha = 0.35f
                            view.parent?.requestDisallowInterceptTouchEvent(true)
                        }
                    }
                    dragStarted
                }
                MotionEvent.ACTION_UP -> {
                    view.parent?.requestDisallowInterceptTouchEvent(false)
                    if (!dragStarted) view.performClick()
                    true
                }
                MotionEvent.ACTION_CANCEL -> {
                    view.parent?.requestDisallowInterceptTouchEvent(false)
                    view.alpha = 1f
                    true
                }
                else -> dragStarted
            }
        }
    }

    private fun placeByTap(number: Int, sourceView: View) {
        if (questionLocked || isGameOver || sourceView.visibility != View.VISIBLE) return
        val nextIndex = placedNumbers.indexOfFirst { it == null }
        if (nextIndex >= 0 && number == currentExercise.orderedNumbers[nextIndex]) {
            placeNumber(nextIndex, number, sourceView)
        } else {
            showSequenceRetry(sourceView)
        }
    }

    private fun placeNumber(targetIndex: Int, number: Int, sourceView: View) {
        if (questionLocked || placedNumbers[targetIndex] != null || sourceView.visibility != View.VISIBLE) return
        placedNumbers[targetIndex] = number
        fillTargetView(targetViews[targetIndex], targetIndex, number)
        sourceView.visibility = View.INVISIBLE

        if (placedNumbers.all { it != null }) {
            questionLocked = true
            celebrateQuestion(targetViews[targetIndex], tvFeedback, successMessage(), ::loadNextQuestion)
        }
    }

    private fun successMessage(): String = getString(
        if (currentExercise.direction == SortDirection.ASCENDING) {
            R.string.sequence_success_small
        } else {
            R.string.sequence_success_big
        }
    )

    private fun fillTargetView(targetView: FrameLayout, index: Int, number: Int) {
        targetView.setOnDragListener(null)
        targetView.findViewById<TextView>(R.id.balloonText).text =
            getString(R.string.number_value, number)
        targetView.findViewById<ImageView>(R.id.balloonBg).imageTintList =
            ColorStateList.valueOf(colorForNumber(number))
        targetView.contentDescription = getString(
            R.string.position_filled,
            index + 1,
            currentExercise.options.size,
            number
        )
    }

    private fun showSequenceRetry(anchor: View) {
        val message = getString(
            if (currentExercise.direction == SortDirection.ASCENDING) {
                R.string.sequence_retry_small
            } else {
                R.string.sequence_retry_big
            }
        )
        showRetryFeedback(anchor, tvFeedback, message)
    }

    private fun colorForNumber(number: Int): Int {
        val colors = listOf(
            R.color.game_blue,
            R.color.game_orange,
            R.color.game_purple,
            R.color.game_green,
            R.color.game_balloon_pink,
            R.color.game_star
        )
        val index = currentExercise.options.indexOf(number).coerceAtLeast(0)
        return ContextCompat.getColor(this, colors[index % colors.size])
    }

    override fun onSaveInstanceState(outState: Bundle) {
        if (::currentExercise.isInitialized) {
            outState.putIntArray(STATE_OPTIONS, currentExercise.options.toIntArray())
            outState.putInt(STATE_DIRECTION, currentExercise.direction.ordinal)
            outState.putIntArray(
                STATE_PLACED,
                placedNumbers.map { it ?: EMPTY_POSITION }.toIntArray()
            )
        }
        outState.putString(STATE_LAST_SIGNATURE, lastSignature)
        outState.putBoolean(STATE_QUESTION_LOCKED, questionLocked)
        super.onSaveInstanceState(outState)
    }

    companion object {
        private const val EMPTY_POSITION = -1
        private const val STATE_OPTIONS = "sequence_options"
        private const val STATE_DIRECTION = "sequence_direction"
        private const val STATE_PLACED = "sequence_placed"
        private const val STATE_LAST_SIGNATURE = "sequence_last_signature"
        private const val STATE_QUESTION_LOCKED = "sequence_question_locked"
    }
}
